tutorials/getting_started/semantic_parsing.md

```diff
@@ -0,0 +1,315 @@
+previous attended input (which is also used as part of the input to the next timestep, for
+"attention feeding").  And because we compute attention over the encoded input representations in
+the LSTM decoder, we also include those inputs (and their mask) as part of the `RnnStatelet`.
+
+`GrammarStatelet` keeps track of the current state of the grammar during decoding: what
+non-terminal is at the top of the non-terminal stack, what actions are available, and so on:
+
+```python
+class GrammarStatelet:
+    def __init__(self,
+                 nonterminal_stack: List[str],
+                 valid_actions: Dict[str, ActionRepresentation],
+                 is_nonterminal: Callable[[str], bool]) -> None:
+        ...
+
+    def is_finished(self) -> bool:
+        ...
+
+    def get_valid_actions(self) -> ActionRepresentation:
+        ...
+
+    def take_action(self, production_rule: str) -> 'GrammarStatelet':
+        ...
+```
+
+To group these things together, we have a `GrammarBasedState` class, that you can actually use with
+a `TransitionFunction`:
+
+```python
+class GrammarBasedState():
+    def __init__(self,
+                 batch_indices: List[int],
+                 action_history: List[List[int]],
+                 score: List[torch.Tensor],
+                 rnn_state: List[RnnStatelet],
+                 grammar_state: List[GrammarStatelet],
+                 possible_actions: List[List[ProductionRuleArray]],
+                 extras: List[Any] = None,
+                 debug_info: List = None) -> None:
+```
+
+As you can see, this keeps track of the basic things that any `State` needs (lists of batch
+indices, action histories, and scores), along with the `RnnStatelets` and `GrammarStatelets`
+discussed above, and a few other fields that are helpful during decoding.
+
+## Defining a language and an execution engine
+
+The previous two sections of this tutorial dealt with a general architecture for training state
+machines, which can be used for any transition system, not just a semantic parser.  For the rest of
+this tutorial, we'll focus on using those state machines to do semantic parsing, by defining a
+language to parse into, an execution engine for that language, and actually executing logical forms
+or programs in a specific context.
+
+And, I'll warn you ahead of time, the way we currently define languages is a bit convoluted and
+difficult to work with.  Sorry.  We're working on a better way to do this, but this is what we have
+for now.
+
+There are currently two ways that you can define a language to parse into with our framework: you
+can use an `nltk`-based logic system, seen in `semparse.transition_functions`, or you can use
+`parsimonious` to write a context-free grammar.  With the `nltk`-based system, you define the
+functions that are available in your (lisp-like) language, with their type signatures, and our code
+builds a grammar for you that can parse statements in that language.  With the `parsimonious`
```

Comment:
For non-lisp like languages, you should mention that we typically use `Parsimonious` because we want to parse into a simplified subset of the full grammar, rather than just using the grammar definition itself, as well as the fact that we need to extract action sequences.